monai/apps/deepgrow/transforms.py

```diff
@@ -0,0 +1,623 @@
+https://github.com/Project-MONAI/MONAI/wiki/MONAI_Design
+"""
+import json
+from typing import Optional, Sequence, Union
+
+import numpy as np
+
+from monai.config import KeysCollection
+from monai.transforms import Resize, SpatialCrop
+from monai.transforms.compose import MapTransform, Randomizable, Transform
+from monai.transforms.spatial.dictionary import InterpolateModeSequence
+from monai.transforms.utils import generate_spatial_bounding_box
+from monai.utils import InterpolateMode, ensure_tuple_rep, min_version, optional_import
+
+measure, _ = optional_import("skimage.measure", "0.14.2", min_version)
+distance_transform_cdt, _ = optional_import("scipy.ndimage.morphology", name="distance_transform_cdt")
+gaussian_filter, _ = optional_import("scipy.ndimage", name="gaussian_filter")
+
+
+class FindAllValidSlicesd(Transform):
+    def __init__(self, label="label", sids="sids"):
+        self.label = label
+        self.sids = sids
+
+    def _apply(self, label):
+        if len(label.shape) != 4:  # only for 3D
+            return None
+
+        sids = []
+        for sid in range(label.shape[1]):  # Assume channel is first
+            if np.sum(label[0][sid]) == 0:
+                continue
+            sids.append(sid)
+        return np.asarray(sids)
+
+    def __call__(self, data):
+        sids = self._apply(data[self.label])
+        if sids is not None and len(sids):
+            data[self.sids] = sids
+        return data
+
+
+class AddInitialSeedPointd(Randomizable, Transform):
+    def __init__(self, label="label", guidance="guidance", sids="sids", sid="sid", connected_regions=6):
+        self.label = label
+        self.sids = sids
+        self.sid = sid
+        self.guidance = guidance
+        self.connected_regions = connected_regions
+
+    def randomize(self, data=None):
+        pass
+
+    def _apply(self, label, sid):
+        dimensions = 3 if len(label.shape) > 3 else 2
+        default_guidance = [-1] * (dimensions + 1)
+
+        dims = dimensions
+        if sid is not None and dimensions == 3:
+            dims = 2
+            label = label[0][sid][np.newaxis]  # Assume channel is first
+
+        label = (label > 0.5).astype(np.float32)
+        blobs_labels = measure.label(label.astype(int), background=0) if dims == 2 else label
+        assert np.max(blobs_labels) > 0, "Not a valid Label"
+
+        pos_guidance = []
+        for ridx in range(1, 2 if dims == 3 else self.connected_regions):
+            if dims == 2:
+                label = (blobs_labels == ridx).astype(np.float32)
+                if np.sum(label) == 0:
+                    pos_guidance.append(default_guidance)
+                    continue
+
+            distance = distance_transform_cdt(label).flatten()
+            probability = np.exp(distance) - 1.0
+
+            idx = np.where(label.flatten() > 0)[0]
+            seed = self.R.choice(idx, size=1, p=probability[idx] / np.sum(probability[idx]))
```

Comment:
random behavior should be put inside ``randomize`` function

Reply:
fixed